contributors/design-proposals/storage/volume-topology-scheduling.md

```diff
@@ -266,41 +304,329 @@ const (
+capacity that has already been provisioned. A missing map entry for a topology
+domain will be treated as zero capacity.
+
+A scheduler predicate will use the reported capacity to determine if dynamic
+provisioning is possible for a given node, taking into account already
+provisioned capacity.
+
+Downsides of this approach include:
+* Multiple topology keys is not supported for capacity reporting. We have not
+seen such use cases so far.
+* There is no common mechanism to report capacity. Each provisioner has to
+update their own fields in the StorageClass if necessary. CSI could potentially
+address this.
+
+#### Local Volume Example
+This example specifies that node1 has 100Gi, and node2 has 200Gi available
+for dynamic provisioning. All other nodes have 0Gi available for dynamic
+provisioning.
+```
+apiVersion: storage.k8s.io/v1
+kind: StorageClass
+metadata:
+  name: standard
+...
+status:
+  capacity:
+    topologyKey: kubernetes.io/hostname
+    capacities:
+      node1: 100Gi
+      node2: 200Gi
+```
+
+TODO: should this be in the local volume spec instead?
+Some downsides specific to the local volume implementation:
+* StorageClass object size can be big for large clusters. Assuming that
+  each node entry in the Capacities map could take up around 100 bytes
+  (size of node name + resource.Quantity), a 1000 node cluster that uses
+  dynamic provisioning of local volumes could add 100Ki to the StorageClass
+  object.
+* If each node has a local volume provisioner, there can be many
+  StorageClass update conflicts if many try to update their capacity at the
+  same time.
+* A security concern is that each local volume provsioner could potentially
+  modify the capacity reported by another node. This could cause
+  provisioning to fail if a node does not have enough capacity, or it could
+  cause the scheduler to skip nodes due to no capacity.
+
+
+#### Relationship to StorageClassQuota
+Although both the new StorageClassCapacity field and existing StorageClassQuota
+manage storage consumption, they address different use cases.
+
+The StorageClassCapacity field is actual available capacity per topology domain
+reported by the storage system, while the StorageClassQuota represents policies
+set by the cluster admin and does not take into account topology domain.
+
+For example, a storage system is configured to support 200Gi in zone1 for a
+StorageClass that is shared by multiple users.  An administrator can use
+StorageClassQuota to restrict capacity usage for each user.
+
+
+## Feature Gates
+PersistentVolume.NodeAffinity and StorageClas.VolumeBindingMode fields will be
+controlled by the VolumeScheduling feature gate, and must be configured in the
+kube-scheduler, kube-controller-manager, and all kubelets.
+
+StorageClass.AllowedTopology and StorageClassCapacity fields will be controlled
+by the DynamicProvisioningScheduling feature gate, and must be configured in the
```

Comment:
Seem the two are not completely independent, maybe we should control both binding and provisioning with a same Feature Gate?

If we enable `VolumeScheduling` and disable `DynamicProvisioningScheduling`, then what's the expected behavior in pv controller? for PVCs of mode `WaitForFirstConsumer`, skip binding but exec provisioning directly?

Reply:
Challenge is static local PVs is beta, and dynamic provisioning will be alpha, so we can't have one feature gate control both.  I see VolumeScheduling as a superset, so if that is disabled, then the delayed dynamic provisioning is also disabled.  Once delayed dynamic provisioning goes to beta, we can combine the two feature gates.